temporalio/ext/src/worker.rs

```diff
@@ -0,0 +1,407 @@
+use std::{cell::RefCell, sync::Arc, time::Duration};
+
+use crate::{
+    client::Client,
+    enter_sync, error, id, new_error,
+    runtime::{AsyncCommand, RuntimeHandle},
+    util::{AsyncCallback, Struct},
+    ROOT_MOD,
+};
+use futures::StreamExt;
+use futures::{future, stream};
+use magnus::{
+    class, function, method, prelude::*, typed_data, DataTypeFunctions, Error, IntoValue, RArray,
+    RString, RTypedData, Ruby, TypedData, Value,
+};
+use prost::Message;
+use temporal_sdk_core::{
+    ResourceBasedSlotsOptions, ResourceBasedSlotsOptionsBuilder, ResourceSlotOptions,
+    SlotSupplierOptions, TunerHolder, TunerHolderOptionsBuilder, WorkerConfigBuilder,
+};
+use temporal_sdk_core_api::errors::PollActivityError;
+use temporal_sdk_core_protos::coresdk::{ActivityHeartbeat, ActivityTaskCompletion};
+
+pub fn init(ruby: &Ruby) -> Result<(), Error> {
+    let class = ruby
+        .get_inner(&ROOT_MOD)
+        .define_class("Worker", class::object())?;
+    class.define_singleton_method("new", function!(Worker::new, 2))?;
+    class.define_singleton_method("async_poll_all", function!(Worker::async_poll_all, 2))?;
+    class.define_singleton_method(
+        "async_finalize_all",
+        function!(Worker::async_finalize_all, 2),
+    )?;
+    class.define_method("async_validate", method!(Worker::async_validate, 1))?;
+    class.define_method(
+        "async_complete_activity_task",
+        method!(Worker::async_complete_activity_task, 2),
+    )?;
+    class.define_method(
+        "record_activity_heartbeat",
+        method!(Worker::record_activity_heartbeat, 1),
+    )?;
+    class.define_method("replace_client", method!(Worker::replace_client, 1))?;
+    class.define_method("initiate_shutdown", method!(Worker::initiate_shutdown, 0))?;
+    Ok(())
+}
+
+#[derive(DataTypeFunctions, TypedData)]
+#[magnus(class = "Temporalio::Internal::Bridge::Worker", free_immediately)]
+pub struct Worker {
+    // This needs to be a RefCell of an Option of an Arc because we need to
+    // mutably take it out of the option at finalize time but we don't have
+    // a mutable reference of self at that time.
+    core: RefCell<Option<Arc<temporal_sdk_core::Worker>>>,
```

Comment:
It looks like the `workers` array you store them in is already protected by using mutexes from the Ruby side.

Rather than doing this, it would make sense to me to store `RefCell<Worker>`s in that array, so you can use mutable references here when finalizing and avoid the option dance

Reply:
So here is the thinking:

* RefCell - needed so I can have a mutable reference
* Option - needed because I need to _take_ the worker out and obtain ownership at finalize time and leave nothing in its place (because I don't own the outer object)
* Arc - Because I need to access this in multiple Rust threads

Can you explain a bit more about removing the Option and Arc parts? How do I get the ability to take the Arc out and unwrap the Arc as the final reference otherwise? Also, how do I have this for use across Rust threads without Arc?

Reply:
What I mean is rather than storing an array of these workers here: https://github.com/temporalio/sdk-ruby/pull/161/files/49d4fc6c62f0ce15d5e3b8f1add3b524de1657a4#diff-3d43a321618f2737c54966b8cad9758e44e5919da7f5bc656c0a3f196ef682ccR5

You could instead store an array of `RefCell<Arc<Worker>>`s which would allow you to use `into_inner()` on the ref cell when finalizing, because you can remove them from the array. That way you can call finalize with `self` instead of a ref, I think.

Reply:
The user provides that worker array at runtime, we don't know it ahead of time. This is part of the ``Worker.run_all` support because Ruby doesn't have a good way to run multiple things concurrently in the standard library without eating threads.

Reply:
Yes, that's more or less what I'm suggesting. That way you can own every worker in the group when it's time to shut down and avoid needing option unwrapping for the more frequently used methods.

Reply:
I don't have it yet, but workers still need to be individually owned to do things like update the client. I can't really move it to the group.